clang-format style fixes

amoudgl opened the pull request:
I have fixed all the style issues in this PR using `clang-format` tool. `.clang-format` file from PR #981 has been used as style guide, which configures all the style options. Please review and let me know how this looks. I'll update the `.clang-format` file accordingly. Thanks!

mlpack-jenkins commented:
Can one of the admins verify this patch?

rcurtin commented on src/mlpack/core/data/binarize.hpp:
```diff
@@ -45,10 +45,10 @@ void Binarize(const arma::Mat<T>& input,
+  T* outPtr = output.memptr();
 
-  #pragma omp parallel for
+#pragma omp parallel for
```
I think it's fine to indent the preprocessor macros, I don't think they should be always at 0 indentation.

amoudgl replied:
There is no option for intending preprocessor separately in `.clang-format` file. It just finds statements with `#` and intends accordingly. One thing we can do is add comments `clang-format off` and `clang-format on` in this particular code section?

rcurtin commented on src/mlpack/core/data/dataset_mapper.hpp:
```diff
@@ -121,8 +119,8 @@ class DatasetMapper
-    ar & data::CreateNVP(types, "types");
-    ar & data::CreateNVP(maps, "maps");
+    ar& data::CreateNVP(types, "types");
+    ar& data::CreateNVP(maps, "maps");
```
This looks like an error, it should be `ar & data::CreateNVP(...)`.

amoudgl replied:
I don't know about this in detail but is it not pass by reference? Because clang-format simply treats it as pass by reference and aligns `&` with the datatype returned.

rcurtin replied:
This is a use of `operator&()`, not a reference.

rcurtin commented on src/mlpack/core/data/dataset_mapper_impl.hpp:
```diff
@@ -21,17 +21,16 @@ namespace data {
-inline DatasetMapper<PolicyType>::DatasetMapper(const size_t dimensionality) :
-    types(dimensionality, Datatype::numeric)
+inline DatasetMapper<PolicyType>::DatasetMapper(const size_t dimensionality)
+    : types(dimensionality, Datatype::numeric)
```
Everything in mlpack has the `:` on the preceding line so that the initialization lists can line up at four-space indentation, so we should move that back to how it was.

amoudgl replied:
Yes, this is an issue in `clang-format` configuration. We have an option for configuring initializer list in clang-format [style guide](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) i.e. `BreakConstructorInitializersBeforeComma`. I have already set it to`false` which matches with mlpack style guide but here I guess length of constructor is exceeding more than 80 chars, so it intends `:` to next line, not so sure why.

rcurtin commented on src/mlpack/core/data/dataset_mapper_impl.hpp:
```diff
@@ -133,15 +130,14 @@ inline Datatype DatasetMapper<PolicyType>::Type(const size_t dimension) const
 {
-  if (dimension >= types.size())
-    types.resize(dimension + 1, Datatype::numeric);
+  if (dimension >= types.size()) types.resize(dimension + 1, Datatype::numeric);
```
I don't like one-line ifs, I think they are harder to read because they don't currently appear anywhere.  Can we change this back to the way it was?

amoudgl replied:
Yes, fixed this by setting `AllowShortIfStatementsOnASingleLine` in the configuration file to false.

rcurtin replied:
Great, thanks!

rcurtin commented on src/mlpack/core/data/load_arff_impl.hpp:
```diff
@@ -94,13 +93,12 @@ void LoadARFF(const std::string& filename,
       {
         throw std::runtime_error("unknown ARFF annotation '" + (*tok.begin()) +
-            "'");
+                                 "'");
```
I have seen this a few times now... if the following line is double-indented, it's fine; it doesn't need to line up with the previous opening brace.

amoudgl replied:
There is an option for this `AlignAfterOpenBracket` in the style guide. Three possible values are: 

1. `Align`: Align parameters on the open bracket, e.g.:
   ```c++
   someLongFunction(argument1,
                    argument2);
   ```
2. `DontAlign`:  Don’t align, instead use 
   ContinuationIndentWidth, e.g.:
   ```c++
   someLongFunction(argument1,
       argument2);
   ```
3. `AlwaysBreak`: Always break after an open bracket, if the parameters don’t fit on a single line, e.g.:
   ```c++
   someLongFunction(
       argument1, argument2);
   ```

I can handle this by setting this configuration option `AlignAfterOpenBracket` to `DontAlign` but it will change the alignment in constructor initializer lists and function declarations too with 4 indent width. Should I go ahead with this?

rcurtin commented on src/mlpack/core/data/load_csv.hpp:
```diff
@@ -114,16 +114,15 @@ class LoadCSV
         // In this case we must pass everything we parse to the MapPolicy.
-        auto firstPassMap = [&](const iter_type& iter)
-        {
+        auto firstPassMap = [&](const iter_type& iter) {
```
The opening brace should be on the following line here, not moved to be on the same line.

amoudgl replied:
I couldn't find any configuration option till now to handle such cases. All the braces alignment have been correctly configured as far as I know with opening braces on a new line. Figuring out any way to handle this.

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -127,7 +124,8 @@ bool Load(const std::string& filename,
         Log::Warn << "'" << filename << "' is comma-separated, not "
-            "tab-separated!" << std::endl;
+                                        "tab-separated!"
+                  << std::endl;
```
This is a strange change, I think it was fine before.

rcurtin commented on src/mlpack/core/data/load_impl.hpp:
```diff
@@ -231,32 +229,34 @@ bool Load(const std::string& filename,
     if (fatal)
-      Log::Fatal << "Attempted to load '" << filename << "' as HDF5 data, but "
+      Log::Fatal
+          << "Attempted to load '" << filename << "' as HDF5 data, but "
```
I think this also was fine before, and looks a bit weird now.

rcurtin commented on src/mlpack/core/data/load_vec_impl.hpp:
```diff
@@ -70,24 +68,22 @@ bool Load(const std::string& filename,
     // It's loaded as a column vector.  We can call the move constructor
     // directly.
-    *((arma::Mat<eT>*) &vec) = std::move(tmp);
+    *((arma::Mat<eT>*)&vec) = std::move(tmp);
```
I think the space should still be there between the cast and the argument.

amoudgl replied:
`SpaceAfterCStyleCast` handles this in clang-format. It has two values `true` or `false`. 

If `true`, a space is inserted after C style casts.
```c++
true:                                  false:
(int)i;                        vs.     (int) i;
```
I have set to `false` but it looks a bug in clang-format for dealing with such complex cases (not sure).

rcurtin replied:
Hmm, maybe if we can show a simple reproducible example we can file a bug upstream.

rcurtin commented on src/mlpack/core/data/map_policies/missing_policy.hpp:
```diff
@@ -85,7 +85,8 @@ class MissingPolicy
   {
-    static_assert(std::numeric_limits<T>::has_quiet_NaN == true,
+    static_assert(
+        std::numeric_limits<T>::has_quiet_NaN == true,
```
Here I also don't see a reason to put the condition on a second line, I think it would have been fine on the first.

rcurtin commented on src/mlpack/core/data/serialization_shim.hpp:
```diff
@@ -123,13 +135,13 @@ inline FirstShim<T> CreateNVP(
+    CreateNVP(T& t,
+              const std::string& name,
+              typename std::enable_if_t<!HasSerialize<T>::value>* = 0,
+              typename std::enable_if_t<!std::is_pointer<T>::value>* = 0)
```
I know this is a weird section of the code, but I think the changes to indentation here aren't necessary; it should still have zero indentation in the function definition.

rcurtin replied:
I think adding `clang-format` comments all over the code just for the sake of formatting is really confusing and we should definitely try to avoid that.  But I know that makes the clang-format configuration process harder... sorry about that... :(

amoudgl replied:
Yes I agree. :) Need to some more hard work on this but I will try my best to come up with some decent solution.

rcurtin commented on src/mlpack/core/data/serialization_shim.hpp:
```diff
@@ -35,37 +35,49 @@ struct HasSerializeFunction
+template<typename T>
+struct SecondNormalArrayShim;
+template<typename T>
+struct PointerShim;
```
No need to split the lines here in my opinion.

amoudgl replied:
There is an option `AlwaysBreakTemplateDeclarations` to handle this. It has two options `true` or `false`. 

```c++
true:                                  false:
template <typename T>          vs.     template <typename T> class C {};
class C {};
```
In most of the mlpack codebase, I have observed that there is always break after template declaration statements. So I set it to `true`, hence this change is there. If I change it to `false`, all the template declaration statements will change. What should we do?

rcurtin commented on src/mlpack/core/data/serialization_shim.hpp:
```diff
@@ -278,8 +292,10 @@ template<typename T>
+  FirstNormalArrayShim(T* t, const size_t len, const std::string& name)
+      : t(t), len(len), name(name)
+  {
+  }
```
I think it's also fine to have an empty function as `{ }` on the line of the definition.

amoudgl commented:
Hi Ryan,

Thanks a lot for taking out time and reviewing this in detail. I have addressed most of your comments, I am still figuring out the rest. Could you please take a look at them?

rcurtin commented:
I went looking through the clang-format documentation, and I see that really there is not a way to have clang-format "ignore" certain parts of the code.  So I have to wonder, is clang-format the right tool for the job here?  Or should we instead be looking at something more flexible (if something else exists)?

amoudgl commented:
We can ignore certain parts of the code by adding comments `clang-format off` and `clang-format on` before and after. But we can't ignore any configuration options e.g. there is a configuration option `AlwaysBreakTemplateDeclarations` which can be set as `true` or `false` but if we want that clang-format should ignore all the template declarations leaving them untouched, we can't do it.

amoudgl commented:
I did some basic search on how to ignore certain checks in clang-format but I couldn't find any way to do that. However, I found that in `clang-tidy` we can pass the checks we want to do and ignore the unnecessary ones. Should I try it?

rcurtin commented:
If we are going to use cpplint instead of clang-format I will go ahead and close this PR.
